File: src/main/java/org/codinjutsu/tools/nosql/mongo/view/model/MongoTreeModelFactory.java
package org.codinjutsu.tools.nosql.mongo.view.model;

import com.mongodb.BasicDBList;
import com.mongodb.BasicDBObject;
import com.mongodb.DBObject;
import org.codinjutsu.tools.nosql.commons.model.SearchResult;
import org.codinjutsu.tools.nosql.commons.view.NoSqlTreeNode;
import org.codinjutsu.tools.nosql.commons.view.nodedescriptor.NodeDescriptor;
import org.codinjutsu.tools.nosql.commons.view.nodedescriptor.NodeDescriptorFactory;
import org.codinjutsu.tools.nosql.mongo.model.MongoResult;
import org.codinjutsu.tools.nosql.mongo.view.nodedescriptor.MongoKeyValueDescriptor;
import org.codinjutsu.tools.nosql.mongo.view.nodedescriptor.MongoResultDescriptor;
import org.codinjutsu.tools.nosql.mongo.view.nodedescriptor.MongoValueDescriptor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.tree.TreeNode;
import java.util.Enumeration;
import java.util.List;

public class MongoTreeModelFactory implements NodeDescriptorFactory<DBObject> {

    @NotNull
    @Override
    public NodeDescriptor createResultDescriptor(@NotNull SearchResult searchResult) {
        return new MongoResultDescriptor();
    }

    @NotNull
    @Override
    public NodeDescriptor createKeyValueDescriptor(@NotNull String key, @Nullable Object value) {
        return MongoKeyValueDescriptor.createDescriptor(key, value);
    }

    @NotNull
    @Override
    public NodeDescriptor createValueDescriptor(int index, @NotNull Object value) {
        return MongoValueDescriptor.createDescriptor(index, value);
    }

    @NotNull
    // todo
    public TreeNode buildJsonTree(@NotNull MongoResult mongoResult) {
        NoSqlTreeNode rootNode = new NoSqlTreeNode(new MongoResultDescriptor(mongoResult.getName()));

        List<DBObject> mongoObjects = mongoResult.getMongoObjects();
        int i = 0;
        for (DBObject mongoObject : mongoObjects) {
            if (mongoObject instanceof BasicDBList) {
                processDbObject(rootNode, mongoObject);
            } else if (mongoObject instanceof BasicDBObject) {//dead code?
                NoSqlTreeNode currentNode = new NoSqlTreeNode(MongoValueDescriptor.createDescriptor(i++, mongoObject));
                processDbObject(currentNode, mongoObject);
                rootNode.add(currentNode);
            }
        }
        return rootNode;
    }

    @Override
    public void processObject(@NotNull NoSqlTreeNode parentNode, @Nullable Object value) {
        if (value instanceof DBObject) {
            processDbObject(parentNode, (DBObject) value);
        }
    }

    void processDbObject(@NotNull NoSqlTreeNode parentNode, DBObject mongoObject) {
        if (mongoObject instanceof BasicDBList) {
            BasicDBList mongoObjectList = (BasicDBList) mongoObject;
            for (int i = 0; i < mongoObjectList.size(); i++) {
                Object mongoObjectOfList = mongoObjectList.get(i);
                NoSqlTreeNode currentNode = new NoSqlTreeNode(MongoValueDescriptor.createDescriptor(i, mongoObjectOfList));
                if (mongoObjectOfList instanceof DBObject) {
                    processDbObject(currentNode, (DBObject) mongoObjectOfList);
                }
                parentNode.add(currentNode);
            }
        } else if (mongoObject instanceof BasicDBObject) {
            BasicDBObject basicDBObject = (BasicDBObject) mongoObject;
            for (String key : basicDBObject.keySet()) {
                Object value = basicDBObject.get(key);
                NoSqlTreeNode currentNode = new NoSqlTreeNode(MongoKeyValueDescriptor.createDescriptor(key, value));
                if (value instanceof DBObject) {
                    processDbObject(currentNode, (DBObject) value);
                }
                parentNode.add(currentNode);
            }
        }
    }

    @NotNull
    @Override
    public DBObject buildDBObject(@NotNull NoSqlTreeNode rootNode) {
        BasicDBObject basicDBObject = new BasicDBObject();
        Enumeration children = rootNode.children();
        while (children.hasMoreElements()) {
            NoSqlTreeNode node = (NoSqlTreeNode) children.nextElement();
            MongoKeyValueDescriptor descriptor = (MongoKeyValueDescriptor) node.getDescriptor();
            Object value = descriptor.getValue();
            if (value instanceof DBObject) {
                if (value instanceof BasicDBList) {
                    basicDBObject.put(descriptor.getKey(), buildDBList(node));
                } else {
                    basicDBObject.put(descriptor.getKey(), buildDBObject(node));
                }
            } else {
                basicDBObject.put(descriptor.getKey(), value);
            }
        }

        return basicDBObject;
    }

    private DBObject buildDBList(NoSqlTreeNode parentNode) {
        BasicDBList basicDBList = new BasicDBList();
        Enumeration children = parentNode.children();
        while (children.hasMoreElements()) {
            NoSqlTreeNode node = (NoSqlTreeNode) children.nextElement();
            MongoValueDescriptor descriptor = (MongoValueDescriptor) node.getDescriptor();
            Object value = descriptor.getValue();
            if (value instanceof DBObject) {
                if (value instanceof BasicDBList) {
                    basicDBList.add(buildDBList(node));
                } else {
                    basicDBList.add(buildDBObject(node));
                }
            } else {
                basicDBList.add(value);
            }
        }
        return basicDBList;
    }
}

File: src/main/java/org/codinjutsu/tools/nosql/mongo/view/model/MongoTreeModelFactory.kt
package org.codinjutsu.tools.nosql.mongo.view.model

import com.mongodb.BasicDBList
import com.mongodb.BasicDBObject
import com.mongodb.DBObject
import org.bson.types.ObjectId
import org.codinjutsu.tools.nosql.commons.style.StyleAttributesProvider.getBooleanAttribute
import org.codinjutsu.tools.nosql.commons.style.StyleAttributesProvider.getNumberAttribute
import org.codinjutsu.tools.nosql.commons.style.StyleAttributesProvider.getObjectAttribute
import org.codinjutsu.tools.nosql.commons.style.StyleAttributesProvider.getObjectIdAttribute
import org.codinjutsu.tools.nosql.commons.style.StyleAttributesProvider.getStringAttribute
import org.codinjutsu.tools.nosql.commons.view.NoSqlTreeNode
import org.codinjutsu.tools.nosql.commons.view.nodedescriptor.NodeDescriptorFactory
import org.codinjutsu.tools.nosql.commons.view.nodedescriptor.keyvalue.BooleanKeyValueDescriptor
import org.codinjutsu.tools.nosql.commons.view.nodedescriptor.keyvalue.DateKeyValueDescriptor
import org.codinjutsu.tools.nosql.commons.view.nodedescriptor.keyvalue.DefaultKeyValueDescriptor
import org.codinjutsu.tools.nosql.commons.view.nodedescriptor.keyvalue.NullKeyValueDescriptor
import org.codinjutsu.tools.nosql.commons.view.nodedescriptor.keyvalue.NumberKeyValueDescriptor
import org.codinjutsu.tools.nosql.commons.view.nodedescriptor.keyvalue.StringKeyValueDescriptor
import org.codinjutsu.tools.nosql.commons.view.nodedescriptor.keyvalue.TypedKeyValueDescriptor
import org.codinjutsu.tools.nosql.commons.view.nodedescriptor.result.StandardResultDescriptor
import org.codinjutsu.tools.nosql.commons.view.nodedescriptor.value.StandardDateValueDescriptor
import org.codinjutsu.tools.nosql.commons.view.nodedescriptor.value.StandardNullValueDescriptor
import org.codinjutsu.tools.nosql.commons.view.nodedescriptor.value.StandardStringValueDescriptor
import org.codinjutsu.tools.nosql.commons.view.nodedescriptor.value.StandardValueDescriptor
import org.codinjutsu.tools.nosql.commons.view.wrapper.ObjectWrapper
import org.codinjutsu.tools.nosql.mongo.model.MongoSearchResult
import org.codinjutsu.tools.nosql.mongo.view.wrapper.MongoObjectWrapper
import java.util.*
import javax.swing.tree.TreeNode

class MongoTreeModelFactory : NodeDescriptorFactory<DBObject> {
    override fun createResultDescriptor(name: String) = StandardResultDescriptor()

    override fun createKeyValueDescriptor(key: String, value: Any?) =
            when (value) {
                null -> NullKeyValueDescriptor(key)
                is Boolean -> BooleanKeyValueDescriptor(key, value.toString().toBoolean())
                is Int -> NumberKeyValueDescriptor(key, value.toString().toInt())
                is Double -> object : TypedKeyValueDescriptor<Double>(key, value, getNumberAttribute()) {
                    override fun setValue(value: Any?) {
                        super.setValue(value.toString().toDouble())
                    }
                }
                is Long -> object : TypedKeyValueDescriptor<Long>(key, value, getNumberAttribute()) {
                    override fun setValue(value: Any?) {
                        super.setValue(value.toString().toLong())
                    }
                }
                is String -> StringKeyValueDescriptor(key, value)
                is Date -> DateKeyValueDescriptor(key, value)
                is ObjectId -> TypedKeyValueDescriptor(key, value, getObjectIdAttribute())
                else -> if (value is DBObject) {
                    TypedKeyValueDescriptor(key, value, getObjectAttribute())
                } else DefaultKeyValueDescriptor(key, value, getStringAttribute())
            }

    override fun createValueDescriptor(index: Int, value: Any): StandardValueDescriptor<*> {
        if (value == null) {
            return StandardNullValueDescriptor(index)
        }

        return when (value) {
            is String -> StandardStringValueDescriptor(index, value)
            is Boolean -> object : StandardValueDescriptor<Boolean>(index, value, getBooleanAttribute()) {
                override fun setValue(value: Any) {
                    super.setValue(value.toString().toBoolean())
                }
            }
            is Number -> object : StandardValueDescriptor<Number>(index, value, getNumberAttribute()) {
                override fun setValue(value: Any) {
                    var number = value as String
                    number = number.replace("(.*)\\..*".toRegex(), "$1")
                    super.setValue(number.toInt())
                }
            }
            is Date -> StandardDateValueDescriptor(index, value)
            is DBObject -> StandardValueDescriptor(index, value, getObjectAttribute())
            else -> StandardValueDescriptor(index, value, getStringAttribute())
        }
    }

    // todo
    fun buildJsonTree(mongoSearchResult: MongoSearchResult): TreeNode {
        val rootNode = NoSqlTreeNode(StandardResultDescriptor(mongoSearchResult.name))

        val mongoObjects = mongoSearchResult.mongoObjects
        var i = 0
        for (mongoObject in mongoObjects) {
            if (mongoObject is BasicDBList) {
                processDbObject(rootNode, mongoObject)
            } else if (mongoObject is BasicDBObject) {//dead code?
                val currentNode = NoSqlTreeNode(createValueDescriptor(i++, mongoObject))
                processDbObject(currentNode, mongoObject)
                rootNode.add(currentNode)
            }
        }
        return rootNode
    }

    override fun processObject(parentNode: NoSqlTreeNode, value: Any?) {
        if (value is DBObject) {
            processDbObject(parentNode, value)
        }
    }

    fun processDbObject(parentNode: NoSqlTreeNode, mongoObject: DBObject) {
        if (mongoObject is BasicDBList) {
            for (i in mongoObject.indices) {
                val mongoObjectOfList = mongoObject[i]
                val currentNode = NoSqlTreeNode(createValueDescriptor(i, mongoObjectOfList))
                if (mongoObjectOfList is DBObject) {
                    processDbObject(currentNode, mongoObjectOfList)
                }
                parentNode.add(currentNode)
            }
        } else if (mongoObject is BasicDBObject) {
            for (key in mongoObject.keys) {
                val value = mongoObject.get(key)
                val currentNode = NoSqlTreeNode(createKeyValueDescriptor(key, value))
                if (value is DBObject) {
                    processDbObject(currentNode, value)
                }
                parentNode.add(currentNode)
            }
        }
    }

    override fun buildDBObject(rootNode: NoSqlTreeNode): DBObject {
        val basicDBObject = BasicDBObject()
        val children = rootNode.children()
        while (children.hasMoreElements()) {
            val node = children.nextElement() as NoSqlTreeNode
            val descriptor = node.descriptor as TypedKeyValueDescriptor
            val value = descriptor.value
            if (value is DBObject) {
                if (value is BasicDBList) {
                    basicDBObject[descriptor.key] = buildDBList(node)
                } else {
                    basicDBObject[descriptor.key] = buildDBObject(node)
                }
            } else {
                basicDBObject[descriptor.key] = value
            }
        }

        return basicDBObject
    }

    private fun buildDBList(parentNode: NoSqlTreeNode): DBObject {
        val basicDBList = BasicDBList()
        val children = parentNode.children()
        while (children.hasMoreElements()) {
            val node = children.nextElement() as NoSqlTreeNode
            val value = node.descriptor.value
            if (value is DBObject) {
                if (value is BasicDBList) {
                    basicDBList.add(buildDBList(node))
                } else {
                    basicDBList.add(buildDBObject(node))
                }
            } else {
                basicDBList.add(value)
            }
        }
        return basicDBList
    }

    override fun isArray(value: Any?) = value is BasicDBList

    override fun toArray(value: Any) = (value as BasicDBList).iterator()

    override fun isObject(value: Any?) = value is DBObject

    override fun createObjectWrapper(value: Any?): ObjectWrapper = MongoObjectWrapper(value as DBObject)
}
